### PR TITLE
Spiderling ai improvements

### PR DESCRIPTION
**Pull request by breadhunt**
## About The Pull Request

Various fixes and tweaks for widow.  Fixes https://github.com/tgstation/TerraGov-Marine-Corps/pull/12728. Fixes  https://github.com/tgstation/TerraGov-Marine-Corps/pull/13712.

Not really a fix to https://github.com/tgstation/TerraGov-Marine-Corps/issues/12981 but Attach Spiderlings should now wake them up from this state, so... yay?

NOW WITH A FIX FOR RESTING SPIDERLINGS!

## Why It's Good For The Game

Widow is a really fun caste. It's a shame it's plagued with bugs, AI constantly breaking.. etc.
Hopefully this PR should fix a lot of those, it seemed to play a lot smoother in local anyway.
## Changelog
:cl:
fix: Spiderlings will no longer wander aimlessly after destroying something, instead returning to the widow. This was big, spiderlings should be a lot more responsive now
add: Attach Spiderlings will now command spiderlings to follow you
fix: This should fix a lot of AI bugginess, especially regarding the Mark Spiderlings ability
fix: Commanding spiderlings to follow you now actually commands spiderlings to follow you
fix: Missing spiderling mark no longer gives spiderlings a midlife crisis, instead just cancelling the ability
fix: Fixed spiderlings de-syncing with their parent widow's rest mode
add: If spiderling mark is targeted at a turf/weeds, and there is a human on that turf, the mark will now target that human. Targeting anything else is unchanged. 
add: If a widow goes crit, instead of fluoride staring at their beloved mother bleeding out, its spiderlings will now attempt to shield it from harm by latching on to its body and physically covering up its sprite
add: If spiderlings are on the same time as widow, any friendly clicks (grabs/abilities/etc) will pass through to the widow. So don't worry, spiderling guard won't block you from saving your widow buddies
fix: Humans can no longer unbuckle a widow's spiderlings from them
add: Spiderlings now turn red when enraged, and blue when guarding
imageadd: Spiderling red and blue colours
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/castes/widow/abilities_widow.dm [hunk 1]
```diff
@@ -188,13 +188,21 @@
...
-	if(!isturf(A))
+	if(!isturf(A) && !istype(A, /obj/alien/weeds))
 		owner.balloon_alert(owner, "Spiderlings attacking " + A.name)
 		succeed_activate()
 	else
-		A = null
-		owner.balloon_alert(owner, "Nothing to attack")
-		fail_activate()
+		for(var/item in A) //Autoaim at humans if weeds or turfs are clicked
+			if(ishuman(item))
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/widow/abilities_widow.dm:196]: `if(!ishuman(item))` then continue

File: code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm [hunk 2]
```diff
@@ -65,10 +64,11 @@
 		go_to_obj_target(source, A)
 		return
 
-/// Sets escorted atom to our pre-defined default escorted atom, which by default is this spiderling's widow
+/// Sets escorted atom to our pre-defined default escorted atom, which by default is this spiderling's widow, and commands the spiderling to follow it
 /datum/ai_behavior/spiderling/proc/only_set_escorted_atom(source, atom/A)
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm:68]: Could you rename this proc to be more descriptive? Something like `revert_to_default_escort`. Also the `atom/A` param is unused so delete that as well.

File: code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm [hunk 3]
```diff
@@ -44,14 +44,13 @@
...
 	RegisterSignals(escorted_atom, list(COMSIG_XENOMORPH_REST, COMSIG_XENOMORPH_UNREST), PROC_REF(toggle_rest))
 	RegisterSignal(escorted_atom, COMSIG_SPIDERLING_MARK, PROC_REF(decide_mark))
+	RegisterSignal(escorted_atom, COMSIG_SPIDERLING_RETURN, PROC_REF(only_set_escorted_atom))
 
 /// Decides what to do when widow uses spiderling mark ability
 /datum/ai_behavior/spiderling/proc/decide_mark(source, atom/A)
 	SIGNAL_HANDLER
 	if(!A)
 		only_set_escorted_atom()
-		UnregisterSignal(atom_to_walk_to, list(COMSIG_MOB_DEATH, COMSIG_QDELETING))
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm:53]: You sure removing this won't cause issues?

**Review by ivanmixo — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm (hunk 3 above)
Comment (breadhunt) [reply code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm:53]: https://www.youtube.com/watch?v=SM69WsJ8r9k

File: code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm [hunk 4]
```diff
@@ -15,43 +21,67 @@
...
+		if(spiderling_state == SPIDERLING_GUARDING)
+			icon_state = "[icon_state] Guarding"
 
 /mob/living/carbon/xenomorph/spiderling/on_death()
 	///We QDEL them as cleanup and preventing them from being sold
 	QDEL_IN(src, TIME_TO_DISSOLVE)
 	return ..()
 
+//If we're covering our widow, any clicks should be transferred to them
+/mob/living/carbon/xenomorph/spiderling/Click(location, control, params)
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm:48]: Click shouldn't be overriden. Try `attack_alien` instead

File: code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm [hunk 5]
```diff
@@ -126,8 +157,50 @@
...
+	SIGNAL_HANDLER
+	INVOKE_ASYNC(src, PROC_REF(revert_to_default_escort))
+	guarding_status = SPIDERLING_NOT_GUARDING
+	var/mob/living/carbon/xenomorph/spiderling/X = mob_parent
+	X.spiderling_state = SPIDERLING_NORMAL
+	X.update_icons()
+
+/datum/ai_behavior/spiderling/ai_do_move()
+	if(guarding_status == SPIDERLING_ATTEMPTING_GUARD)
+		if(get_dist(mob_parent, atom_to_walk_to) <= 1)
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm:178]: &&

**Review by ivanmixo — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm (hunk 3 above)
Comment (ivanmixo) [reply code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm:53]: Guess we'll see in the TM. Also don't self-resolve stuff that isn't resolved.

File: code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm (hunk 4 above)
Comment (breadhunt) [reply code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm:48]: attack_alien won't cover for ability usage, e.g heals. afaik, the only way to cover for that is through either a catch-all click() or overriding more specific stuff like right/middle/alt clicks() - would that be acceptable? are there any other, cleaner methods?

File: code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm (hunk 3 above)
Comment (breadhunt) [reply code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm:53]: woops! got it.

File: code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm [hunk 6]
```diff
@@ -15,43 +21,67 @@
...
 
 /mob/living/carbon/xenomorph/spiderling/on_death()
 	///We QDEL them as cleanup and preventing them from being sold
 	QDEL_IN(src, TIME_TO_DISSOLVE)
 	return ..()
 
+//If we're covering our widow, any clicks should be transferred to them
+/mob/living/carbon/xenomorph/spiderling/attack_alien(mob/living/carbon/xenomorph/X, damage_amount, damage_type, damage_flag, effects, armor_penetration, isrightclick)
+	if(!get_dist(src, spidermother))
+		if(isxeno(x))
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm:50]:
```suggestion
	if(!get_dist(src, spidermother)) && if(isxeno(X))
```

File: code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm [hunk 7]
```diff
@@ -15,43 +21,67 @@
 	allow_pass_flags = PASS_XENO
 	pass_flags = PASS_XENO|PASS_LOW_STRUCTURE
 	density = FALSE
+	// The widow that this spiderling belongs to
+	var/mob/living/carbon/xenomorph/spidermother
+	// What sprite state this - normal, enraged, guarding? Used for update_icons()
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm:26]: Triple slashes for autodoc

File: code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm [hunk 8]
```diff
@@ -15,43 +21,67 @@
...
+			icon_state = "[icon_state] Enraged"
+		if(spiderling_state == SPIDERLING_GUARDING)
+			icon_state = "[icon_state] Guarding"
 
 /mob/living/carbon/xenomorph/spiderling/on_death()
 	///We QDEL them as cleanup and preventing them from being sold
 	QDEL_IN(src, TIME_TO_DISSOLVE)
 	return ..()
 
+//If we're covering our widow, any clicks should be transferred to them
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/spiderling/spiderling.dm:47]: Triple slashes here too

File: code/datums/components/riding/riding_mob.dm [hunk 9]
```diff
@@ -313,3 +325,23 @@
 	for(var/mob/living/rider AS in carrying_widow.buckled_mobs)
 		carrying_widow.unbuckle_mob(rider)
 		REMOVE_TRAIT(rider, TRAIT_IMMOBILE, WIDOW_ABILITY_TRAIT)
+
+// Spiderlings latch on to crit widows when guarding and cannot be kicked off..
+/datum/component/riding/creature/widow/ride_check(mob/living/rider)
+	var/mob/living/widow = parent
+	if(widow.stat == UNCONSCIOUS)
```
Comment (BraveMole) [inline code/datums/components/riding/riding_mob.dm:332]: return widow.stat == UNCONSCIOUS

**Review by BraveMole — APPROVED** [review]

**Review by BraveMole — APPROVED** [review]

**Pull request merged**